### DISCUSSION TITLE
3D frictional contact does not converge.

### DISCUSSION THREAD

**Discussion by dzheng26** [idaholab/moose#17387] [Q&A Modules: Solid mechanics] [answered]
New to Moose. I tried to have a 3D frictional contact simulation. It is a tire-like tube push against a rigid surface. The rigid surface is modeled as one element with specified movement at all nodes.

Without friction (mono04-cnt.i), the example works fine. However it does not converge with friction (mono04-frc.i). If I change the frictional coefficient to zero (mono04-frc0.i), it works like the frictionless case.

Could someone help me on this example?

The test data is attached.
[test_data.tar.gz](https://github.com/idaholab/moose/files/6172241/test_data.tar.gz)

**Comment by dzheng26** [comment, marked as answer]
Also in "modules/contact/examples/3d_berkovich", the file "indenter.e" used by the input file is missing.

**Reply by jiangwen84** [reply]
@recuero Could you add the missing file? Thanks.

**Reply by recuero** [reply]
There is a journal file in the repository that can be used to generate the exodus file. The mesh is quite large. I am attaching it here, remove the txt extension.

[indenter.e.txt](https://github.com/idaholab/moose/files/6172769/indenter.e.txt)

**Reply by dzheng26** [reply]
Thanks @recuero . A small README file in the folder can be helpful.

**Reply by recuero** [reply]
I am not sure if you are referring to the journal file conversion to Exodus. But, in case you haven't come across it: https://mooseframework.inl.gov/modules/contact/BerkovichIndenterNodeFace.html

**Reply by dzheng26** [reply]
Thanks for sending me the .e file. Sorry for not make a follow up of this early. I tested the example 3d_berkovich. I have to make a small changed of the input file since the "boty111" is not defined on the .e file and print_perf_log has to be replaced. The example converged but slowly (I did not ran it all the way to the end). There are many non-linear iterations for each time step. 

Follow the example, I also made my small example converged. However it also converged slowly. Much too slow in my opinion and to other code I tried.

Input file changes
`<   [./boty111]
<     type = DirichletBC
<     variable = disp_y
<     boundary = 111
<     value = 0.0
<   [../]
<   [./botx111]
<     type = DirichletBC
<     variable = disp_x
<     boundary = 111
<     value = 0.0
<   [../]
< 
219c206
<   print_perf_log = true
---
>   perf_graph = true
`
Convergence example
`Time Step 90, time = 0.36, dt = 0.004
 0 Nonlinear |R| = 2.298919e-02
 1 Nonlinear |R| = 2.035932e-03
 2 Nonlinear |R| = 1.890794e-03
 3 Nonlinear |R| = 1.748625e-03
 4 Nonlinear |R| = 1.504827e-03
 5 Nonlinear |R| = 1.277548e-03
 6 Nonlinear |R| = 1.046155e-03
 7 Nonlinear |R| = 8.971936e-04
 8 Nonlinear |R| = 7.813298e-04
 9 Nonlinear |R| = 7.658471e-04
10 Nonlinear |R| = 6.122786e-04
11 Nonlinear |R| = 1.882083e-04
12 Nonlinear |R| = 1.696259e-04
13 Nonlinear |R| = 1.302233e-04
14 Nonlinear |R| = 1.204901e-04
15 Nonlinear |R| = 1.975728e-07
`

**Reply by recuero** [reply]
Problems with friction tend to need a larger number of nonlinear iterations. In addition, if you are using the slip damper, it will probably be adding to the number of nonlinear iterations by controlling the slip state of the nodes in contact. 

I assume it's your small problem that you are comparing to other codes. To improve convergence in MOOSE, I'd recommend investigating or tuning input parameters, such as `normal_smoothing_method`, `normal_smoothing_distance`, and `penalty`.

**Reply by dzheng26** [reply]
@recuero , thanks for the hints. I will have another trial. BTW, which formulation is best for 3D frictional contact? I use "penalty" currently.

**Reply by recuero** [reply]
For frictional contact, `tangential_penalty` usually yields good results.

**Reply by dzheng26** [reply]
@recuero Following your help, I did more tests. Here is what I found.
First `normal_smoothing_method` is not allowed in my compiled code although from peacock I found `edge_based` and `nodal_normal_based` two options. So I cannot test it. I tested the `penalty` and `tangential_penalty` formulations with `penalty` swap (0.01, 0.1, 1.0 and 10.0) and `normal_smoothing_distance` swap (0.01, 0.1, 1.0 and 10.0). The cases for `penalty=10` did not converge. All the others converged. Here is the total number of Nonlinear iterations for each test cases.

For penalty method (P0_01SD0_1 means penalty=0.01 and normal_smoothing_distance=0.1) 
```P0_01SD0_01.log:| 91
P0_01SD0_1.log:| 95
P0_01SD1_0.log:| 102
P0_01SD10_0.log:| 150

P0_1SD0_01.log:| 333
P0_1SD0_1.log:| 353
P0_1SD1_0.log:| 359
P0_1SD10_0.log:| 334

P1_0SD0_01.log:| 3120
P1_0SD0_1.log:| 2843
P1_0SD1_0.log:| 2748
P1_0SD10_0.log:| 6558
```

For tangential_penalty method
```
TP0_01SD0_01.log:| 106
TP0_01SD0_1.log:| 102
TP0_01SD1_0.log:| 98
TP0_01SD10_0.log:| 105

TP0_1SD0_01.log:| 239
TP0_1SD0_1.log:| 488
TP0_1SD1_0.log:| 239
TP0_1SD10_0.log:| 242

TP1_0SD0_01.log:| 918
TP1_0SD0_1.log:| 993
TP1_0SD1_0.log:| 944
TP1_0SD10_0.log:| 953
```
The `tangential_penalty` method convergences better.

When checking the results, there are visually noticeable penetrations through the contact plane even with `penalty=1.0` for `penalty` formulation. WIth `tangential_penalty` method, there is no (little) penetration by formulation. However, there are visually noticeable violation of the friction law for `penalty=0.01 or 0.1`. It is hard to see for `penalty=1.0` cases.

With the best usable setting (tangential_penalty with penalty=1.0 and normal_smoothing_distance=1.0), the number of nonlinear iterations is quite high (944 for 21 load steps). IMO, there are rooms for improvement for the contact algorithms.

**Reply by recuero** [reply]
The convergence behavior and interpenetration values depend heavily on the selection of parameters. "Penalties" are usually chosen in relation to the stiffnesses of the bodies involved. The normal smoothing distance is given in parametric coordinates, so it should range from 0 to 0.5. But this parameter's effect depends on the problem surfaces.
